Proposal: Use `blkid` to detect fs type of device instead of `lsblk`.

cofyc opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/devel/pull-requests.md#the-pr-submit-process and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/devel/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/devel/pull-requests.md#write-release-notes-if-needed
4. If the PR is unfinished, see how to mark it: https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md#marking-unfinished-pull-requests
-->

**What this PR does / why we need it**:

Use `blkid` to detect fs type of device instead of `lsblk`. 

`lsblk` depends on `udev`, and device driver's udev rules. If udev rules are not installed properly, `lsblk` could not get fs type of disk. This will cause problems, e.g. expanding volume will fail because it could not detect fs type of disk.

Take `rbd`  as example, `lsblk -n -o FSTYPE /dev/rbd0` command actually read fs type from `/run/udev/data/b252:0` (may vary on different machines, see real file from `strace lsblk -n -o FSTYPE /dev/rbd0` ).

This file `/run/udev/data/b252:0` is generated by `udev` and device driver's udev files. If machine don't have `/lib/udev/rules.d/60-ceph-by-parttypeuuid.rules` udev rule file installed (this file is from `ceph-osd` package on ubuntu), it will not be properly generated, e.g:

```
# cat /run/udev/data/b251:0
S:rbd/<pool>/<image>
I:13234059587579
E:ID_FS_TYPE=
E:net.ifnames=0
G:systemd
```

`lsblk -n -o FSTYPE /dev/rbd0` will get empty fs type.

[60-ceph-by-parttypeuuid.rules](https://github.com/ceph/ceph/blob/v13.0.0/udev/60-ceph-by-parttypeuuid.rules) is udev rule, which underlyingly runs `blkid` commands to get infos of device, then store them in udev file.

If we use `blkid` to get fs type, kubelet volume manager will not depend on `udev` stuffs. Currently, if kubelet node does not have `60-ceph-by-parttypeuuid.rules` installed (from `ceph-osd` package), it will fail to get fs type of rbd image. Even administrator install `ceph-osd` later, it can not get fs type of previous mapped images (udev data files not be to updated automatically).

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

`udevadm test` logs:

- Ubuntu 16.04 (ceph-common installed), without ceph-osd: [without-ceph-osd.log](https://github.com/kubernetes/kubernetes/files/1678512/without-ceph-osd.log)
- Ubuntu 16.04 (ceph-common installed), with ceph-osd: [with-ceph-osd.log](https://github.com/kubernetes/kubernetes/files/1678509/with-ceph-osd.log)

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
NONE
```

Commit:
Use `blkid` to get fs type of device.

`lsblk` reads fs type info from udev files. If udev rules are not
installed. `lsblk` could not get correct fs type. This will cause
problems, e.g. expanding volume depends on fs type of disk.

cofyc commented:
/cc @jsafrane @rootfs

gnufied commented:
/ok-to-test

cofyc commented:
/test pull-kubernetes-cross
/test pull-kubernetes-e2e-gce
/test pull-kubernetes-e2e-kops-aws

cofyc commented:
`pull-kubernetes-cross` fails due to missing `modifyHostNetworkOptionForContainer` func  (https://github.com/kubernetes/kubernetes/pull/58174#issuecomment-361834731).

cofyc commented:
/sig storage

cofyc commented:
/test pull-kubernetes-cross

cofyc commented:
`pull-kubernetes-cross` fails due to https://github.com/kubernetes/kubernetes/issues/59223.

jsafrane commented on pkg/util/mount/safe_format_and_mount_test.go:
```diff
@@ -159,30 +159,20 @@ func TestSafeFormatAndMount(t *testing.T) {
-			expectedError: fmt.Errorf("failed to mount the volume as \"ext3\", it already contains unknown data, probably partitions. Mount error: unknown filesystem type '(null)'"),
-		},
-		{
-			description: "Test that 'lsblk' is called and reports empty partition",
```
So, what blkid returns when a device contains a partition? lsblk detected it and kubelet did not reformat the device. Now it reformats it, wiping all user data.

cofyc replied:
@jsafrane 

In addition to print  FS type (by `-s TYPE`), we can print `PTTYPE` too.

`blkid -p -s TYPE -s PTTYPE -o export <device>`:

Example output if device contains parititions:

```
# blkid -p -s TYPE -s PTTYPE -o export /dev/sda1
DEVNAME=/dev/sda
PTTYPE=dos
```

Example output if device does not contains partitions:

```
# blkid -p -s TYPE -s PTTYPE -o export /dev/sda1
DEVNAME=/dev/sda1
TYPE=ext2
```

PTTYPE is partition type, referring to docs: https://www.kernel.org/pub/linux/utils/util-linux/v2.21/libblkid-docs/libblkid-Partitions-probing.html

cofyc replied:
@jsafrane Updated. I removed the last "empty partitions" test case, because we only need to know if disk has partitions or not. Though it's possible to get partition lists from [`PART_ENTRY_*` tags](https://www.kernel.org/pub/linux/utils/util-linux/v2.21/libblkid-docs/libblkid-Partitions-probing.html).

Commit:
Use `blkid` to get fs type of device.

If a parition table type is detected, returns a special non-empty string
as filesystem type.

jsafrane commented:
/assign
I generally do not have anything against `blkid`, however the test failures look suspicious.

cc: @kubernetes/sig-storage-pr-reviews for broader discussion. Is anyone against using blkid instead of lsblk?

rootfs commented:
both are in util-linux package, no an issue to me.

rootfs commented on pkg/util/mount/mount_linux.go:
```diff
@@ -549,23 +549,38 @@ func (mounter *SafeFormatAndMount) GetDiskFormat(disk string) (string, error) {
-	// Split lsblk output into lines. Unformatted devices should contain only
-	// "\n". Beware of "\n\n", that's a device with one empty partition.
-	output = strings.TrimSuffix(output, "\n") // Avoid last empty line
+	var (
```
nit `var fstype, pttype string`

cofyc replied:
Fixed.

rootfs commented on pkg/util/mount/mount_linux.go:
```diff
@@ -549,23 +549,38 @@ func (mounter *SafeFormatAndMount) GetDiskFormat(disk string) (string, error) {
-		// The device is unformatted and has no dependent devices
-		return "", nil
+	if len(pttype) > 0 {
+		glog.V(4).Infof("Disk %s detected parition table type: %s", pttype)
```
partition

cofyc replied:
Fixed.

rootfs commented:
howeever, blkid manpage says otherwise and I do have a problem to run blkid with non privileged id.
```
It  is  recommended  to  use  lsblk(8) command to get information about
       block devices rather than blkid.  lsblk(8) provides  more  information,
       better  control  on output formatting and it does not require root per‐
       missions to get actual information.
```

gnufied commented:
I do not have anything against using `blkid`. I briefly straced invocation of both `blkid` and `lsblk` and output of `blkid` looked good (There was no udev reads).  Having said that - yeah, the test failures look suspicious and hopefully nothing breaks.

afaict - kubelet always runs with privileged permissions, so `blkid` requiring privilged id should not be a blocker.

msau42 commented:
I think we should prerun all the storage e2es for this change.  The pull e2es may not exercise this path.

I can run against COS.

msau42 commented:
/assign

cofyc commented:
/test pull-kubernetes-cross

cofyc commented:
/pull-kubernetes-e2e-gce
/pull-kubernetes-e2e-kops-aws

Commit:
Use `blkid` to get fs type of device.

For `blkid`, if the specified token (TYPE/PTTYPE, etc) was
not found, or no (specified) devices could be identified, an
exit code of 2 is returned.

cofyc commented:
/pull-kubernetes-unit

cofyc commented:
/test pull-kubernetes-unit

cofyc commented:
@jsafrane @gnufied @rootfs @msau42 
hi, all tests are passed now. My mistake, fixed now.

msau42 commented:
@verult ran the e2es and manually tested against GCE PD to make sure the partition detection works

jsafrane commented:
@msau42 @verult  thanks
/lgtm

mykaul commented:
Why not use '--usages filesystem' to filter for file systems in the call to blkid?

cofyc commented:
I didn't know this parameter when I wrote this patch. It's my mistake. I think we can have it if it does not introduce incompatibilities. Done a quick test, it uses less `read` calls.

```
# strace blkid -p -s TYPE -s PTTYPE -o export /dev/sda1   |& grep -i '^read(' | wc -l
29
$ strace blkid -p -u filesystems -s TYPE -s PTTYPE -o export /dev/sda1   |& grep -i '^read(' | wc -l
25
```
